### PR TITLE
Update torch.set_default_dtype doc

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41263 Update torch.set_default_dtype doc**

Differential Revision: [D22482989](https://our.internmc.facebook.com/intern/diff/D22482989)

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
 
     The default floating point dtype is initially ``torch.float32``.
 
     Args:
         d (:class:`torch.dtype`): the floating point dtype to make the default
 
     Example::
-
-        >>> torch.tensor([1.2, 3]).dtype           # initial default for floating point is torch.float32
+        >>> # initial default for floating point is torch.float
```
Comment (gchanan) [inline torch/__init__.py:286]: nit: this isn't a huge deal, but I think the previous wording was slightly better, because it matches what is returned for the next line.  As it is now, people will wonder why `torch.float` and  `torch.float32` don't match exactly.

File: torch/__init__.py [hunk 2]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
+        >>> torch.tensor([1.2, 3]).dtype
         torch.float32
-        >>> torch.set_default_dtype(torch.float64)
-        >>> torch.tensor([1.2, 3]).dtype           # a new floating point tensor
+        >>> # initial default for floating point is torch.cfloat
```
Comment (gchanan) [inline torch/__init__.py:289]: same here, use `torch.complex64`.

File: torch/__init__.py [hunk 3]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
+        >>> torch.tensor([1.2, 3j]).dtype
+        torch.complex64
+        >>> torch.set_default_tensor_type(torch.DoubleTensor)
```
Comment (gchanan) [inline torch/__init__.py:292]: I don't think we want to document this -- the type interface isn't complete (e.g. there's no quantized, XLA) -- it's really logically deprecated.

File: torch/__init__.py [hunk 4]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
 
 
 def set_default_dtype(d):
-    r"""Sets the default floating point dtype to :attr:`d`. This type will be
-    used as default floating point type for type inference in
-    :func:`torch.tensor`.
+    r"""Sets the default floating point dtype to :attr:`d`.
+    This type is also used:
```
Comment (gchanan) [inline torch/__init__.py:274]: this says "also used" but it's sort of defined by how it's used, i.e. "sets the default floating point dtype" doesn't really mean anything outside of how it' s used.  I'd just change this to something like:
"This dtype is used as:"

File: torch/__init__.py [hunk 5]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
+    1. As default floating point type for type inference in :func:`torch.tensor`.
+    2. To determine the default complex dtype. The default complex dtype is set to
+       ``torch.complex128`` if default floating point tensor type is ``torch.float64``,
+       else it's set to ``torch.complex64``
```
Comment (gchanan) [inline torch/__init__.py:278]: nit: else -> otherwise.

File: torch/__init__.py [hunk 6]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
 
 
 def set_default_dtype(d):
-    r"""Sets the default floating point dtype to :attr:`d`. This type will be
-    used as default floating point type for type inference in
-    :func:`torch.tensor`.
+    r"""Sets the default floating point dtype to :attr:`d`.
+    This type is also used:
+    1. As default floating point type for type inference in :func:`torch.tensor`.
```
Comment (gchanan) [inline torch/__init__.py:275]: this could be a little clearer (I know it's coming from the prior writeup).  But "default floating point type" isn't really defined yet.  Maybe something like:
"As the dtype inferred for python floats" (and similar for complex).

File: torch/__init__.py (hunk 5 above)
Comment (gchanan) [reply torch/__init__.py:278]: is it not true that this is the corresponding complex dtype for the float default?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: torch/__init__.py (hunk 1 above)
Comment (anjali411) [reply torch/__init__.py:286]: ahh ok

File: torch/__init__.py (hunk 3 above)
Comment (anjali411) [reply torch/__init__.py:292]: yeah that makes sense: updated to `torch.set_default_dtype(torch.float64)`

File: torch/__init__.py (hunk 5 above)
Comment (anjali411) [reply torch/__init__.py:278]: that's true

File: torch/__init__.py (hunk 6 above)
Comment (anjali411) [reply torch/__init__.py:275]: updated

File: torch/__init__.py (hunk 4 above)
Comment (anjali411) [reply torch/__init__.py]: updated

File: torch/__init__.py [hunk 7]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
 
 
 def set_default_dtype(d):
-    r"""Sets the default floating point dtype to :attr:`d`. This type will be
-    used as default floating point type for type inference in
-    :func:`torch.tensor`.
+    r"""Sets the default floating point dtype to :attr:`d`.
+    This dtype is:
+    1. The inferred dtype for python floats in :func:`torch.tensor`.
+    2. Used to determine the default complex dtype. The default complex dtype is set to
```
Comment (gchanan) [inline torch/__init__.py:276]: same issue with 1 -- this should say something like "the inferred dtype for python complex numbers"

File: torch/__init__.py [hunk 8]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
+       ``torch.complex128`` if default floating point tensor type is ``torch.float64``,
+       otherwise it's set to ``torch.complex64``
```
Comment (gchanan) [inline torch/__init__.py:278]: I'm curious: why did you word it this way instead of saying something about the corresponding float dtype?

Comment (anjali411) [reply torch/__init__.py:278]: The only reason why I didn't it frame it that way is because we don't have a complex dtype corresponding to `torch.bfloat16`, `torch.float16` (technically we do but the support for ComplexHalf is almost non-existing)

File: torch/__init__.py (hunk 7 above)
Comment (anjali411) [reply torch/__init__.py:276]: does this sound better?

```
Used to infer dtype for python complex numbers. The default complex dtype is set to
 ``torch.complex128`` if default floating point dtype is ``torch.float64``, otherwise it's set to ``torch.complex64``
```

Comment (gchanan) [reply torch/__init__.py:276]: lgtm!

**Review by gchanan — APPROVED** [review]

File: torch/__init__.py [hunk 9]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
 
 def set_default_dtype(d):
-    r"""Sets the default floating point dtype to :attr:`d`. This type will be
-    used as default floating point type for type inference in
-    :func:`torch.tensor`.
+    r"""Sets the default floating point dtype to :attr:`d`.
+    This dtype is:
+    1. The inferred dtype for python floats in :func:`torch.tensor`.
+    2. Used to infer dtype for python complex numbers. The default complex dtype is set to
+       ``torch.complex128`` if default floating point dtype is ``torch.float64``,
```
Comment (mruberry) [inline torch/__init__.py:277]: "if the default..."

File: torch/__init__.py [hunk 10]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
 
 
 def set_default_dtype(d):
-    r"""Sets the default floating point dtype to :attr:`d`. This type will be
-    used as default floating point type for type inference in
-    :func:`torch.tensor`.
+    r"""Sets the default floating point dtype to :attr:`d`.
+    This dtype is:
+    1. The inferred dtype for python floats in :func:`torch.tensor`.
```
Comment (mruberry) [inline torch/__init__.py:275]: This isn't quite right since it's the inferred type for Python floats in other contexts, too:

```
torch.set_default_dtype(torch.double)
(t + 5.).dtype
: torch.float64
```

File: torch/__init__.py [hunk 11]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
+    2. Used to infer dtype for python complex numbers. The default complex dtype is set to
```
Comment (mruberry) [inline torch/__init__.py:276]: You may want to make this a note for clarity:

"This setting also determines the inferred dtype of complex numbers. If the default floating point dtype is torch.float64 then complex numbers are inferred to have a dtype of torch.complex128, otherwise they are assumed to have a dtype of torch.complex64."

File: torch/__init__.py [hunk 12]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
+       ``torch.complex128`` if default floating point dtype is ``torch.float64``,
+       otherwise it's set to ``torch.complex64``
 
     The default floating point dtype is initially ``torch.float32``.
```
Comment (mruberry) [inline torch/__init__.py:280]: This might also be noteworthy?

File: torch/__init__.py [hunk 13]
```diff
@@ -270,22 +270,30 @@ def set_default_tensor_type(t):
...
 
     Args:
         d (:class:`torch.dtype`): the floating point dtype to make the default
 
     Example::
-
-        >>> torch.tensor([1.2, 3]).dtype           # initial default for floating point is torch.float32
+        >>> # initial default for floating point is torch.float32
```
Comment (mruberry) [inline torch/__init__.py:286]: Following the above comment, maybe it'd be nice to have an example that isn't `torch.tensor` based?

File: torch/__init__.py (hunk 10 above)
Comment (anjali411) [reply torch/__init__.py:275]: I think this falls under the `torch.tensor` case since when you add a scalar to a tensor, we first make a zero dim tensor containing the scalar (hence the dtype of this new tensor in this case will be `torch.float64`

Comment (mruberry) [reply torch/__init__.py:275]: That's true in the first case but also an implementation detail I wouldn't expect a user to be be aware of. In the second case I don't think `torch.tensor()` is involved?
